node/lib/util/git_util.js

```diff
@@ -307,6 +307,48 @@ exports.fetch = co.wrap(function *(repo, remoteName) {
 });
 
 /**
+ * Fetch the specified `sha` from the "origin" repository of the specifieded
+ * `repo`.
+ *
+ * @async
+ * @param {NodeGit.Repository} repo
+ * @param {String} sha
+ */
+exports.fetchSha  = co.wrap(function *(repo, sha) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isString(sha);
+
+    // First, try to get the commit.  If we succeed, no need to fetch.
+
+    try {
+        yield repo.getCommit(sha);
+        return;                                                       // RETURN
+    }
+    catch (e) {
+    }
+
+    let origin;
+    try {
+        origin = yield repo.getRemote("origin");
+    }
+    catch (e) {
+        throw new UserError(`No origin at ${repo.workdir()}.`);
+    }
+
+    const execString = `\
+git -C ${repo.workdir()} fetch-pack -q '${origin.url()}' ${sha}
+`;
+    try {
+        return yield exec(execString);
+    }
+    catch (e) {
+        throw new UserError(e.message);
```

Comment:
I want to lose the stack trace here. We throw`UserError` specifically because we want to show the error to the user; generally, those should not have stack traces.